src/txreconciliation.h

```diff
@@ -0,0 +1,585 @@
+
+    /**
+     * Used to keep track of ongoing reconciliations (or lack of them) per peer.
+     */
+    Mutex m_states_mutex;
+    std::unordered_map<NodeId, ReconciliationState> m_states GUARDED_BY(m_states_mutex);
+
+    /**
+     * Reconciliation should happen with peers in the same order, because the efficiency gain is the
+     * highest when reconciliation set difference is predictable. This queue is used to maintain the
+     * order of peers chosen for reconciliation.
+     */
+    Mutex m_queue_mutex;
+    std::deque<NodeId> m_queue GUARDED_BY(m_queue_mutex);
+
+    /**
+     * Reconciliations are requested periodically:
+     * every RECON_REQUEST_INTERVAL seconds we pick a peer from the queue.
+     */
+    std::chrono::microseconds m_next_recon_request{0};
+    void UpdateNextReconRequest(std::chrono::microseconds now) EXCLUSIVE_LOCKS_REQUIRED(m_queue_mutex)
+    {
+        m_next_recon_request = now + RECON_REQUEST_INTERVAL / m_queue.size();
+    }
+
+    /**
+     * Used to schedule the next initial response for any pending reconciliation request.
+     * Respond to all requests at the same time to prevent transaction possession leak.
+     */
+    std::chrono::microseconds m_next_recon_respond{0};
+    std::chrono::microseconds NextReconRespond()
+    {
+        auto current_time = GetTime<std::chrono::microseconds>();
+        if (m_next_recon_respond < current_time) {
+            m_next_recon_respond = current_time + RECON_RESPONSE_INTERVAL;
+        }
+        return m_next_recon_respond;
+    }
+
+    public:
+
+    TxReconciliationTracker() {};
+
+    /**
+     * TODO: document
+     */
+    std::tuple<bool, bool, uint32_t, uint64_t> SuggestReconciling(const NodeId peer_id, bool inbound);
+
+    /**
+     * If a peer was previously initiated for reconciliations, get its current reconciliation state.
+     * Note that the returned instance is read-only and modifying it won't alter the actual state.
+     */
+    bool EnableReconciliationSupport(const NodeId peer_id, bool inbound,
+        bool recon_requestor, bool recon_responder, uint32_t recon_version, uint64_t remote_salt,
+        size_t outbound_flooders);
+
+    /**
+     * If a it's time to request a reconciliation from the peer, this function will return the
+     * details of our local state, which should be communicated to the peer so that they better
+     * know what we need.
+     */
+    Optional<std::pair<uint16_t, uint16_t>> MaybeRequestReconciliation(const NodeId peer_id);
+
+    /**
+     * Record an (expected) reconciliation request with parameters to respond when time comes. All
+     * initial reconciliation responses will be done at the same time to prevent privacy leaks.
+     */
+    void HandleReconciliationRequest(const NodeId peer_id, uint16_t peer_recon_set_size, uint16_t peer_q);
+
+    /**
+     * TODO document
+     */
+    Optional<std::vector<uint8_t>> MaybeRespondToReconciliationRequest(const NodeId peer_id);
```

Comment:
What's the difference for the caller between an empty vector and a nullopt here?